doc/man-pages/bas2x11basic.1
.TH BAS2X11BASIC 1 19-Jan-2022 "Version 1.28" "X11-Basic"
.SH NAME
bas2x11basic \- ANSI-Basic to X11-Basic converter
.SH SYNOPSIS
.B bas2x11basic
.RI "[ " option " ] "
[\fB\-o\fR \fIoutfile\fR]
.IR <filename>

.SH DESCRIPTION

X11-Basic is a dialect of the BASIC programming language with graphics.

X11-Basic is a structured programming language. Structure is a form of visual
and functional encapsulation in which multiple line sections of a program  look
and act like single units. The beginning and end of blocks are marked by
descriptive keyword delimiters. Lines within blocks are generally indented to
make the block visible. Most important: X11-Basic does not use line numbers.

.B bas2x11basic
can be used to make X11-Basic template programs out of ANSI basic files. It
helps converting old (real) Basic Programs with line numbers and multiple
commands per line to the X11-Basic structure. Because there are so many
different BASIC versions and dialects around, in most cases you will have to edit
these files produced manually. But most of the work will already have been done
by this converter.

.PP
For a desciption of the X11-Basic language and usage of the interpreter see
.I xbasic(1)\fR.

.SH Filenames

.B bas2x11basic 
processes input files with extention .bas. The default output filename is 
.I b.bas\fR.

.SH OPTIONS
.TP
.BR \-o " " \fIfile\fR
Place output in file 
.I file
\&. 
If you do not specify `\|\c
.B \-o\c
\&\|', the default is to use
`\|\c
.B b.bas\c
\&\|'
.TP
.BR \-\-version
shows version, program and other information and exits
.TP
.BR \-h ", " \-\-help
print a short help.
.PP
For detailed help and description of the implemented basic commands take a look
at the README file. 


.SH EXAMPLES
.nf
bas2x11basic ansibasic.bas
    produces an output file b.bas

bas2x11basic ansibasic.bas -o newname.bas
    produces an output file newname.bas
.fi

You can find some samples of former ANSI basic programs in 
.I /usr/share/doc/packages/X11Basic/examples.

.SH The ANSI Basic to X11-Basic converter

Actually this converter is not yet complete. Further improvenents may be applied
in future. Please let me know about ANSI-Basic programs which will not be
converted.

.SH VERSION
Page was created for V.1.09

.SH BUG REPORTS 

If you find a bug in the ANSI Basic to X11-Basic converter, you should report
it. But first, you should make sure that it really is a bug, and that it appears
in the latest version of the X11-Basic package that you have.

Once you have determined that a bug actually exists, mail a bug report to
kollo@users.sourceforge.net. If you have a fix, you are welcome to mail that as
well! Suggestions may be mailed to the X11-Basic bug tracking system.

Comments and bug reports concerning this manual page should be directed to
kollo@users.sourceforge.net.

.SH AUTHOR
Markus Hoffmann <kollo@users.sourceforge.net>

.SH COPYRIGHT
Copyright (C) 1997-2022 Markus Hoffmann <kollo@users.sourceforge.net>

This program is free software; you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free Software
Foundation; either version 2 of the License, or (at your option) any later
version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE. See the GNU General Public License for more details.

.SH SEE ALSO
x11basic(1), rsc2gui(1)


doc/HISTORY

Version history of X11-Basic
============================

In the beginning there were absurd plans, long sleepless nights, and incredible
boredom.

Very First Version (1990)
--------------------------
It was a Mailbox-Program with Basic features (only character-Graphics with
Vt100EMU) on my ATARI ST (!).

Version 0.1 First version with a version number    (1997)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
- Mailbox-Program was rewritten in c on my Linux-PC
- first X11-Commands

Version 0.2 Alpha (1998)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
- improved X11-Basic for use at the physics institute in Bonn
- Fixed last Problems of the Parser (didn't like -1*(4-3) )
- Added more X-features (redraw)
- First portation to a HP-Workstation
- Readline library in use

Version 1.00 Beta (Mar 1999)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~
- Makefile and configure script adapted from STonX
- True-color Displays
- X rootwindow, multiple windows
- INKEY$
- First GPL-Distribution
- FILESELECT,ALERT

Version 1.01 (Jun 1999)
~~~~~~~~~~~~~~~~~~~~~~
- POINT()

Version 1.02 (Jun 2000)
~~~~~~~~~~~~~~~~~~~~~~
- rsrc_load
- shell, daemon

Version 1.03 (Mar 2001)
~~~~~~~~~~~~~~~~~~~~~~~
- copyarea, sget, sput, get,put
- fill-Patterns
- drop down menues
- String-Variables may now contain nul-characters.
- POKE,DPOKE,LPOKE,EVEN(),ODD()
- LINK,UNLINK
- CVI(),CVD(),CVF(),CVL()
- MKI$(),MKD$(),MKL$(),MKF$()
- ON x GOSUB proc1,proc2,...
- ON x GOTO label1,label2,...
- INSTR(),RINSTR()

Version 1.04 (Apr. 2001)
~~~~~~~~~~~~~~~~~~~~~~~~
- splitted off shared library libx11basic.so (300kB)
- bug fixed which caused print a,b to segmentation fault
- DUMP "#"
- SYM_ADR()
- CALL, EXEC, EXEC()  with parameter list
- DUMP "K" DUMP "F"

Version 1.05 (Aug. 2001)
~~~~~~~~~~~~~~~~~~~~~~~~
- PRINT USING
- SYSTEM$()
- GLOB()
- RBOX, PRBOX
- EDIT, SAVE
- HELP
- new system errors
- started to translate everything into english

Version 1.06 (Jan. 2002)
~~~~~~~~~~~~~~~~~~~~~~~~

- included serial device support
- improved FILESELECT
- modified ALERT
- improved COLOR
- FORM_DIAL(), FORM_DO(), OBJC_DRAW()
- GPRINT

Version 1.07 (Aug. 2002)
~~~~~~~~~~~~~~~~~~~~~~~~
- Fehler mit Exponentialdarstellungen von Zahlen mit negativem Exponent
  ist nun endlich behoben !!!
- Anbindung an tine-Bibliothek und Kontrollsystem ist nun moeglich

Version 1.08 (Jan. 2003)
~~~~~~~~~~~~~~~~~~~~~~~~
- Support for tine-library
- a first MS-WINDOWS version is now available
- some new functions
- FIT and FIT_LINEAR
- SORT command
There was a bug introduced in the MENU command. Fixed it in release 1.08-4.

Version 1.09 (June 2003)
~~~~~~~~~~~~~~~~~~~~~~~~
- ANSI-Basic to X11-Basic converter included
- New example programs
- updated man-page and documentation
- new SOUND command
- EVAL, EVAL()
- modified serial line support
- modified the configure procedure
- some modifications for portability
- will now compile on FreeBSD and CYGWIN

Version 1.10 (Feb 2004)
~~~~~~~~~~~~~~~~~~~~~~~~
- new REPLACE$() function 
- improved speed 
- bug fixed in ROUND() function
- fixed bug in PUT_BITMAP
- fixed bug in serial device handling
- added sysvars COLS and ROWS
- new INODE/DEVICE/MODE/NLINK/SIZE file functions

Version 1.11 (Oct 2004)
~~~~~~~~~~~~~~~~~~~~~~~~
- Parser now accepts unique short commands
- fixed little bug in INPUT
- new chapters in manual
- new commands: TOPW, BOTTOMW, INFOW, FULLW, CLIP
- found bug in ON ERROR and ON BREAK
- new FILL command
- changed TAB treatement in sourcecode
- worked on TINE implementation
- worked on the ARRAY functions
- new SHL(), SHR() function

Version 1.12 (Mar 2005)
~~~~~~~~~~~~~~~~~~~~~~~~
- fixed bug with DEFMARK
- experimental LISTSELECT
- worked on MATRIX functions: SOLVE(),INV(),DET()
- OPEN "UU"  UDP-Sockets
- CONNECT, SEND, RECEIVE
- New String Functions: REVERSE$(), MTFE$(), MTFD$(), RLD$(), RLE$()
- CRC()
- New Code Engines: BETE$(), BWTD$(), COMPRESS$()

Version 1.13 (Feb 2006)
~~~~~~~~~~~~~~~~~~~~~~~~
- fixed bug in INPUT$()
- woked in the TINE implementation (TINELISTEN, TINEBROADCAST)
- new MKA$(), CVA()
- fixed bug in PAUSE/DELAY
- implemented ioctl() function

Version 1.14 (Apr 2007)
~~~~~~~~~~~~~~~~~~~~~~~~
- improved the init_program routines
- woked on the bytecode compiler
- new programs xbbc and xbvm  + man pages
- the bytecodecompiler and virtual machine are still full of bugs
- some sample programs can already be compiled to bytecode, 
  they run 4 to 10 times faster 
- Bug fixed which caused "Command  not allowed in interactive mode" errors
  This bug was introduced in Version 1.14 since release 1
- worked on the control system routines (DOOCS)
- fixed bug with fgetc
- crosscompiled it with arm-linux for embedded linux
- new CHDIR command
- new SPAWN command (does still not work as wanted)
- created doocsxbasic modification to use the FLASH control system at DESY


Release notes for Version 1.15 (Feb 2008)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
- New command: PIPE
- New commands: GET_SCREENSIZE, GET_GEOMETRY
- Implemented BMP file format (for PUT)
- produced a framebuffer-graphics version of X11-basic


Release notes for Version 1.16 (Feb 2010)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
- Bugfix in SORT (Feb 2010)
- new function word$() June 2010
- Improved DATA/RESTORE handling Dez 2010
- Improved xbbc/bytecode/xbvm
- Fixed memory leak accuring in varptr()
- added xbbc and xbvm to windows distribution
- started xb2c project
- new psydocompiler for MS-WINDOWS, can produce standalone .exe files 
- Bug fixed: LINEINPUT and LINEINPUT$() now can read arbitrary long lines
- fixed bug with PBOX: now 2nd coordinate is inside the filled area.
- new commands WATCH/(UNWATCH)/FILEEVENT$
- fixed bug with bitmaps (PUT)
- added support for bitmaps with transparency/alpha channel
- fixed bug in DEFFILL
- fixed bug in objc_draw()

Release notes for Version 1.17 (May 2011)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
- Adapted windows version to use the SDL library (2011-04)

examples/compiler/bas2x11basic.bas
' Utility to convert ANSI BASIC Programs to X11Basic
'
'* This file is part of X11BASIC, the basic interpreter for Unix/X
'* ======================================================================
'* X11BASIC is free software and comes with NO WARRANTY - read the file
'* COPYING for details
'*
' (c) Markus Hoffmann 2003 (letzte Bearbeitung: 10.12.2002)
' (letzte Bearbeitung: 10.02.2004)
'
'

' 2012-10-29 Modifications by Marcos Cruz (programandala.net)

i=1
' outputfilename$="b.bas" ! xxx original
outputfilename$=@yyyymmddhhmmss$()+"_converted.xbas" ! xxx new, only for debugging
WHILE LEN(PARAM$(i))
  IF LEFT$(PARAM$(i))="-"
    IF PARAM$(i)="--help" OR PARAM$(i)="-h"
      @intro
      @using
    ELSE IF PARAM$(i)="--version"
      @intro
      QUIT
    ELSE IF PARAM$(i)="-o"
      INC i
      IF LEN(PARAM$(i))
        outputfilename$=PARAM$(i)
      ENDIF
    ENDIF
  ELSE
    inputfile$=PARAM$(i)
    IF NOT EXIST(inputfile$)
      PRINT "bas2x11basic: "+inputfile$+": file or path not found"
      CLR inputfile$
    ENDIF
  ENDIF
  INC i
WEND
IF (WIN32? OR ANDROID?) AND (LEN(inputfile$)=0 OR inputfile$="bas2x11basic.bas")
  FILESELECT "Load ANSI basic program","./*.bas","ANSI.bas",inputfile$
ENDIF
IF LEN(inputfile$)
  rumpf$=inputfile$
  WHILE LEN(rumpf$)
    WORT_SEP rumpf$,"/",1,a$,rumpf$
  WEND
  f$=a$
  rumpf$=a$
  WORT_SEP a$,".",1,rumpf$,typ$
  IF typ$="bas"
    IF EXIST(outputfilename$)
      PRINT "bas2x11basic: Outputfilename already exists: ";outputfilename$
    ELSE
      @convert
    ENDIF
  ELSE
    PRINT f$+": file not recognized: File format not recognized"
  ENDIF
ELSE
  PRINT "bas2x11basic: No input files"
ENDIF
QUIT
PROCEDURE intro
  PRINT "ANSI BASIC to X11-Basic Converter V.1.11 (c) Markus Hoffmann 2003-2004"
  VERSION
RETURN
PROCEDURE using
  PRINT "Usage: bas2x11basic [options] file..."
  PRINT "Options:"
  PRINT "  -h, --help		    Display this information"
  PRINT "  -o <file>		    Place the output into <file>"
RETURN
PROCEDURE convert
  DIM linenr(3000),zeile$(3000),labeling(3000),procs(3000)
  anzlabel=0
  anzprocs=0
  anzline=0
  OPEN "O",#2,outputfilename$
  ' xxx original header:
  '  PRINT #2,"' bas2x11basic V.1.10 ("+f$+")"
  '  PRINT #2,"'	   (c) Markus Hoffmann "+date$+" "+time$
  ' xxx new suggested header:
  PRINT #2,"' "+outputfilename$
  PRINT #2,"' ";@iso_time$()
  PRINT #2,"' Automatic conversion of "+f$+" to X11-Basic"
  PRINT #2,"' by bas2x11basic V.1.11 "
  PRINT #2,"' Copyright (C) 2002-2004 Markus Hoffmann"
  PRINT #2
  ' PASS 1
  PRINT "PASS 1"
  pass=1
  OPEN "I",#1,inputfile$
  WHILE NOT EOF(#1)
    LINEINPUT #1,t$
    t$=XTRIM$(t$)
    IF LEN(t$)
      WORT_SEP t$," ",1,b$,t$
      linenr(anzline)=VAL(b$)
      zeile$(anzline)=t$
      INC anzline
      IF UPPER$(LEFT$(t$,2))="IF"
        @processifline(t$)
      ELSE
        @processline(t$)
      ENDIF
    ENDIF
  WEND
  CLOSE #1
  PRINT #2,"' ";anzline;" lines."
  PRINT #2,"' ";anzlabel;" labels."
  PRINT #2,"' ";anzprocs;" procs."
  PRINT #2,"' ----- Start of program -----"
  ' PASS 2
  PRINT "PASS 2"
  pass=2

  FOR i=0 TO anzline-1
    flag=0
    lln=linenr(i)
    FOR j=0 TO anzlabel-1
      IF labeling(j)=lln
        flag=1
      ENDIF
    NEXT j
    IF flag
      PRINT #2,"L";lln;":"
    ENDIF
    flag=0
    lln=linenr(i)
    FOR j=0 TO anzprocs-1
      IF procs(j)=lln
        flag=1
      ENDIF
    NEXT j
    IF flag
      PRINT #2,"PROCEDURE P";lln
    ENDIF
    tt$=TRIM$(zeile$(i))
    IF UPPER$(LEFT$(tt$,2))="IF"
      @processifline(tt$)
    ELSE
      @processline(tt$)
    ENDIF
  NEXT i
  PRINT #2,"' ----- End of program -----"
  CLOSE #2
RETURN

PROCEDURE processline(t$)
  t$=REPLACE$(t$,", ",",")
  WHILE LEN(t$)
    WORT_SEP t$,":",1,b$,t$
    b$=XTRIM$(b$)
    b$=REPLACE$(b$," REM "," ! ")
    b$=REPLACE$(b$," = ","=")
    b$=REPLACE$(b$,"= ","=")
    b$=REPLACE$(b$," + ","+")
    b$=REPLACE$(b$," - ","-")
    b$=REPLACE$(b$," * ","*")
    b$=REPLACE$(b$," / ","/")
    b$=REPLACE$(b$," ^ ","^")
    b$=REPLACE$(b$," > ",">")
    b$=REPLACE$(b$," < ","<")
    b$=REPLACE$(b$," <> ","<>")
    @processcommand(b$)
  WEND
RETURN
PROCEDURE processifline(t$)
  LOCAL ifauf,a$,b$
  ' xxx why is this in the original?, it seems debugging code:
  ' IF pass<>1
  '   PRINT #2,"' "+t$
  ' ENDIF
  t$=REPLACE$(t$,", ",",")
  WHILE LEN(t$)
    WORT_SEP t$,":",1,b$,t$
    b$=XTRIM$(b$)
    b$=REPLACE$(b$," REM "," ! ")
    b$=REPLACE$(b$," = ","=")
    b$=REPLACE$(b$,"= ","=")
    b$=REPLACE$(b$," + ","+")
    b$=REPLACE$(b$," - ","-")
    b$=REPLACE$(b$," * ","*")
    b$=REPLACE$(b$," / ","/")
    b$=REPLACE$(b$," ^ ","^")
    b$=REPLACE$(b$," > ",">")
    b$=REPLACE$(b$," < ","<")
    b$=REPLACE$(b$," <> ","<>")
    WORT_SEP b$," ",1,b$,a$
    IF UPPER$(b$)="IF"
      WORT_SEP a$,"THEN ",1,a$,c$
      IF LEN(c$)=0
        WORT_SEP a$," ",1,a$,c$
      ENDIF
      IF pass<>1
        PRINT #2,"IF "+a$
      ENDIF

      IF LEN(c$)
        WORT_SEP c$,"ELSE ",1,c$,d$
        IF c$=STR$(VAL(c$))
          @processcommand("GOTO "+c$)
        ELSE
          @processcommand(c$)
        ENDIF
        IF LEN(d$)
          IF pass<>1
            PRINT #2,"ELSE"
          ENDIF
          IF d$=STR$(VAL(d$))
            @processcommand("GOTO "+d$)
          ELSE
            @processcommand(d$)
          ENDIF
        ENDIF
        ifauf=1
      ENDIF
    ELSE
      @processcommand(b$+" "+a$)
    ENDIF
  WEND
  IF ifauf
    IF pass<>1
      PRINT #2,"ENDIF"
    ENDIF
  ENDIF
RETURN

PROCEDURE processcommand(b$)
  b$=REPLACE$(b$,"PRINT"+CHR$(34),"PRINT "+CHR$(34))
  b$=REPLACE$(b$,"INPUT"+CHR$(34),"INPUT "+CHR$(34))
  WORT_SEP b$," ",1,b$,a$
  IF UPPER$(b$)="GOTO"
    WORT_SEP a$," ",1,a$,c$
    IF pass=1
      labeling(anzlabel)=VAL(a$)
      INC anzlabel
    ELSE
      PRINT #2,b$;" L";a$+" "+c$
    ENDIF
  ELSE IF UPPER$(b$)="GOSUB"
    WORT_SEP a$," ",1,a$,c$
    IF pass=1
      procs(anzprocs)=VAL(a$)
      INC anzprocs
    ELSE
      PRINT #2,"@P";a$+" "+c$
    ENDIF
  ELSE IF UPPER$(b$)="IF"
    '    PRINT #2,"' ERROR: No if here !" ! xxx original
    @processifline(b$+" "+a$) ! xxx new
  ELSE
    IF pass>1
      PRINT #2,b$;" ";a$
    ENDIF
  ENDIF
RETURN

FUNCTION yyyymmddhhmmss$() ! xxx new, only for debugging
  LOCAL year$,month$,day$,hour$,minute$,second$
  year$=RIGHT$(date$,4)
  month$=LEFT$(date$,2)
  day$=MID$(date$,4,2)
  hour$=LEFT$(time$,2)
  minute$=MID$(time$,4,2)
  second$=RIGHT$(time$,2)
  RETURN year$+month$+day$+hour$+minute$+second$
ENDFUNCTION

FUNCTION iso_time$() ! xxx new, just a suggestion
  LOCAL year$,month$,day$,hour$,minute$,second$
  year$=RIGHT$(date$,4)
  month$=LEFT$(date$,2)
  day$=MID$(date$,4,2)
  hour$=LEFT$(time$,2)
  minute$=MID$(time$,4,2)
  second$=RIGHT$(time$,2)
  RETURN year$+"-"+month$+"-"+day$+" "+hour$+":"+minute$+":"+second$
ENDFUNCTION
